token/program-2022/src/extension.rs

```diff
@@ -1,13 +1,279 @@
+        let length_end_index = length_start_index.saturating_add(pod_get_packed_len::<Length>());
+        let value_start_index = length_end_index;
+
+        let extension_type = ExtensionType::try_from(&tlv_data[start_index..type_end_index])?;
+        // got to an empty spot, can init here, or move forward if not initing
+        if extension_type == ExtensionType::Uninitialized {
+            if init {
+                return Ok(TlvIndices(
+                    start_index,
+                    length_start_index,
+                    value_start_index,
+                ));
+            } else {
+                start_index = length_start_index;
+            }
+        } else if extension_type == V::TYPE {
+            // found an instance of the extension that we're initializing, abort!
+            if init {
+                return Err(ProgramError::InvalidArgument);
+            } else {
+                return Ok(TlvIndices(
+                    start_index,
+                    length_start_index,
+                    value_start_index,
+                ));
+            }
+        } else {
+            let length = pod_from_bytes::<Length>(&tlv_data[length_start_index..length_end_index])?;
+            let value_end_index = value_start_index.saturating_add(usize::from(*length));
+            start_index = value_end_index;
+        }
+    }
+    Err(ProgramError::InvalidAccountData)
+}
+
+fn check_not_multisig(input: &[u8]) -> Result<(), ProgramError> {
+    if input.len() == Multisig::LEN {
+        Err(ProgramError::InvalidAccountData)
+    } else {
+        Ok(())
+    }
+}
+
+fn check_account_type<S: BaseState>(account_type: AccountType) -> Result<(), ProgramError> {
+    if account_type != S::ACCOUNT_TYPE {
+        Err(ProgramError::InvalidAccountData)
+    } else {
+        Ok(())
+    }
+}
+
+fn type_and_tlv_indices_unchecked<S: BaseState>(
+    rest_input: &[u8],
+) -> Result<(usize, usize), ProgramError> {
+    if rest_input.is_empty() {
+        Ok((0, 0))
+    } else {
+        let account_type_index = Account::LEN.saturating_sub(S::LEN);
+        let tlv_start_index = account_type_index.saturating_add(size_of::<AccountType>());
+        Ok((account_type_index, tlv_start_index))
+    }
+}
+
+fn type_and_tlv_indices<S: BaseState>(rest_input: &[u8]) -> Result<(usize, usize), ProgramError> {
+    if rest_input.is_empty() {
+        Ok((0, 0))
+    } else {
+        let type_index = Account::LEN.saturating_sub(S::LEN);
+        let tlv_start_index = type_index.saturating_add(size_of::<AccountType>());
+        Ok((type_index, tlv_start_index))
+    }
+}
+
+/// Encapsulates immutable base state data (mint or account) with possible extensions
+#[derive(Debug, PartialEq)]
+pub struct StateWithExtensions<'data, S: BaseState> {
+    /// Unpacked base data
+    pub base: S,
+    /// Unpacked base data
+    pub account_type: AccountType,
+    /// Slice of data containing all TLV data, deserialized on demand
+    tlv_data: &'data [u8],
```

Comment:
For the `Mint` implementation, it looks like the `account_type` and `tlv_data` will end up immediately after the base Mint.  But this means that the length of a Mint with extensions could be the same as Account::LEN. Rather than adding padding, like I think we discussed earlier, how about extending `check_not_multisig` to be `check_not_multisig_or_account`?

Reply:
Couldn't we then have a situation in which we're not totally sure if an account is an `Account` or a `Mint`?  A mint with extensions that takes it past 165 could be indiscernible from an `Account` with an extension, even if we add the account type. For example, let's say:

```
Account: 165 bytes... + [2, 0, 3, 0, 100, ....]
                         ^     ^       ^     ^
                    acct type  extension length data...
```
```
Mint: 82 bytes... + 83 bytes of random other data + [2, 0, 3, 0, 100, ....]
                                                     ^ data in extension just happens to look like this
```

With the current approach, we only start writing the TLV data after `Account::LEN`, which means we always know that the account type is going to be right after that.  We can special case checking for a Multisig, because those aren't extensible under any circumstances.

Reply:
hmm, yeah I think you're right.  thanks